src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/NodeMonitoringFailoverPlugin.java

```diff
@@ -208,11 +208,18 @@ public void releaseResources() {
     this.next.releaseResources();
   }
 
-  protected void initNodeKeys(Connection connection) {
+  /**
+   * If the connection is using a different hostname, add that host name to the set.
+   *
+   * @param connection The current connection.
+   * @param hostInfo   Information about the current connection.
+   * @return Updated set of node keys.
+   */
+  protected Set<String> getUpdatedNodeKeys(Connection connection, HostInfo hostInfo) {
```

Comment:
Does this have the potential to add duplicated keys if the same connection is used? Would it be better to move nodeKeys.clear() to the top of this method rather than inside failover()?

Reply:
Shouldn't be possible. We are using `Set`, which is [a collection that contains no duplicate elements](https://docs.oracle.com/javase/8/docs/api/java/util/Set.html).